File: apps/backend/src/modules/goals/entities/goal.entity.ts
export enum GoalPaceStatus {
  COMPLETED = 'COMPLETED', // Target reached or marked complete
  AHEAD = 'AHEAD',         // Savings is ahead of expected pace (> 105% of scheduled pace)
  ON_TRACK = 'ON_TRACK',   // Savings is on schedule (90% - 105% of scheduled pace)
  BEHIND = 'BEHIND',       // Savings is behind schedule (< 90% of scheduled pace)
  OVERDUE = 'OVERDUE',     // Target date passed and target not achieved
}

export class GoalEntity {
  id: string;
  userId: string;
  linkedAccountId?: string;
  linkedAccountName?: string;
  name: string;
  targetAmount: number;
  currentAmount: number;
  remainingAmount: number;
  progressPercentage: number;
  targetDate: string;
  daysRemaining: number;
  paceStatus: GoalPaceStatus;
  expectedPaceAmount: number;
  paceDifference: number; // positive = ahead, negative = shortfall
  requiredMonthlySavings: number;
  requiredWeeklySavings: number;
  requiredDailySavings: number;
  isCompleted: boolean;
  createdAt: string;
  updatedAt: string;

  static fromRow(goalRow: any, linkedAccountName?: string, referenceDate: Date = new Date()): GoalEntity {
    const targetAmount = parseFloat(goalRow.target_amount) || 0.0;
    const currentAmount = parseFloat(goalRow.current_amount) || 0.0;
    const isCompleted = goalRow.is_completed || currentAmount >= targetAmount;
    const remainingAmount = Math.max(0, targetAmount - currentAmount);
    const progressPercentage = targetAmount > 0 ? Math.min(100, (currentAmount / targetAmount) * 100) : 0;

    const createdAt = new Date(goalRow.created_at || referenceDate);
    const targetDate = new Date(goalRow.target_date);
    
    // Day calculations
    const msPerDay = 1000 * 60 * 60 * 24;
    const totalDays = Math.max(1, Math.ceil((targetDate.getTime() - createdAt.getTime()) / msPerDay));
    const elapsedDays = Math.max(0, Math.min(totalDays, Math.ceil((referenceDate.getTime() - createdAt.getTime()) / msPerDay)));
    const daysRemaining = Math.max(0, Math.ceil((targetDate.getTime() - referenceDate.getTime()) / msPerDay));

    // Expected pace based on elapsed time vs total time
    const expectedPaceAmount = (elapsedDays / totalDays) * targetAmount;
    const paceDifference = currentAmount - expectedPaceAmount;

    // Determine Pace Status
    let paceStatus: GoalPaceStatus;
    if (isCompleted) {
      paceStatus = GoalPaceStatus.COMPLETED;
    } else if (daysRemaining === 0 && currentAmount < targetAmount) {
      paceStatus = GoalPaceStatus.OVERDUE;
    } else if (expectedPaceAmount <= 0) {
      paceStatus = currentAmount > 0 ? GoalPaceStatus.AHEAD : GoalPaceStatus.ON_TRACK;
    } else if (currentAmount >= expectedPaceAmount * 1.05) {
      paceStatus = GoalPaceStatus.AHEAD;
    } else if (currentAmount >= expectedPaceAmount * 0.9) {
      paceStatus = GoalPaceStatus.ON_TRACK;
    } else {
      paceStatus = GoalPaceStatus.BEHIND;
    }

    // Required rates to reach goal by target date
    const monthsRemaining = Math.max(0.1, daysRemaining / 30.4375);
    const weeksRemaining = Math.max(0.1, daysRemaining / 7);
    const requiredMonthlySavings = remainingAmount > 0 ? remainingAmount / monthsRemaining : 0;
    const requiredWeeklySavings = remainingAmount > 0 ? remainingAmount / weeksRemaining : 0;
    const requiredDailySavings = remainingAmount > 0 ? remainingAmount / Math.max(1, daysRemaining) : 0;

    const entity = new GoalEntity();
    entity.id = goalRow.id;
    entity.userId = goalRow.user_id;
    entity.linkedAccountId = goalRow.linked_account_id;
    entity.linkedAccountName = linkedAccountName || goalRow.accounts?.name;
    entity.name = goalRow.name;
    entity.targetAmount = targetAmount;
    entity.currentAmount = currentAmount;
    entity.remainingAmount = remainingAmount;
    entity.progressPercentage = Math.round(progressPercentage * 100) / 100;
    entity.targetDate = goalRow.target_date;
    entity.daysRemaining = daysRemaining;
    entity.paceStatus = paceStatus;
    entity.expectedPaceAmount = Math.round(expectedPaceAmount * 100) / 100;
    entity.paceDifference = Math.round(paceDifference * 100) / 100;
    entity.requiredMonthlySavings = Math.round(requiredMonthlySavings * 100) / 100;
    entity.requiredWeeklySavings = Math.round(requiredWeeklySavings * 100) / 100;
    entity.requiredDailySavings = Math.round(requiredDailySavings * 100) / 100;
    entity.isCompleted = isCompleted;
    entity.createdAt = goalRow.created_at;
    entity.updatedAt = goalRow.updated_at;

    return entity;
  }
}

File: apps/mobile/lib/features/goals/domain/goal_entity.dart
import 'package:flutter/material.dart';

enum GoalPaceStatus {
  completed,
  ahead,
  onTrack,
  behind,
  overdue;

  static GoalPaceStatus fromString(String val) {
    switch (val.toUpperCase()) {
      case 'COMPLETED':
        return GoalPaceStatus.completed;
      case 'AHEAD':
        return GoalPaceStatus.ahead;
      case 'ON_TRACK':
        return GoalPaceStatus.onTrack;
      case 'BEHIND':
        return GoalPaceStatus.behind;
      case 'OVERDUE':
        return GoalPaceStatus.overdue;
      default:
        return GoalPaceStatus.onTrack;
    }
  }

  String get label {
    switch (this) {
      case GoalPaceStatus.completed:
        return 'Achieved';
      case GoalPaceStatus.ahead:
        return 'Ahead of Schedule';
      case GoalPaceStatus.onTrack:
        return 'On Track';
      case GoalPaceStatus.behind:
        return 'Behind Pace';
      case GoalPaceStatus.overdue:
        return 'Target Date Passed';
    }
  }

  Color get color {
    switch (this) {
      case GoalPaceStatus.completed:
        return const Color(0xFF10B981);
      case GoalPaceStatus.ahead:
        return const Color(0xFF06B6D4);
      case GoalPaceStatus.onTrack:
        return const Color(0xFF10B981);
      case GoalPaceStatus.behind:
        return const Color(0xFFF59E0B);
      case GoalPaceStatus.overdue:
        return const Color(0xFFEF4444);
    }
  }

  IconData get icon {
    switch (this) {
      case GoalPaceStatus.completed:
        return Icons.check_circle_rounded;
      case GoalPaceStatus.ahead:
        return Icons.trending_up_rounded;
      case GoalPaceStatus.onTrack:
        return Icons.done_all_rounded;
      case GoalPaceStatus.behind:
        return Icons.warning_amber_rounded;
      case GoalPaceStatus.overdue:
        return Icons.error_outline_rounded;
    }
  }
}

class GoalEntity {
  final String id;
  final String userId;
  final String? linkedAccountId;
  final String? linkedAccountName;
  final String name;
  final double targetAmount;
  final double currentAmount;
  final double remainingAmount;
  final double progressPercentage;
  final DateTime targetDate;
  final int daysRemaining;
  final GoalPaceStatus paceStatus;
  final double expectedPaceAmount;
  final double paceDifference;
  final double requiredMonthlySavings;
  final double requiredWeeklySavings;
  final double requiredDailySavings;
  final bool isCompleted;
  final DateTime createdAt;
  final DateTime updatedAt;

  const GoalEntity({
    required this.id,
    required this.userId,
    this.linkedAccountId,
    this.linkedAccountName,
    required this.name,
    required this.targetAmount,
    required this.currentAmount,
    required this.remainingAmount,
    required this.progressPercentage,
    required this.targetDate,
    required this.daysRemaining,
    required this.paceStatus,
    required this.expectedPaceAmount,
    required this.paceDifference,
    required this.requiredMonthlySavings,
    required this.requiredWeeklySavings,
    required this.requiredDailySavings,
    required this.isCompleted,
    required this.createdAt,
    required this.updatedAt,
  });

  factory GoalEntity.fromJson(Map<String, dynamic> json) {
    return GoalEntity(
      id: json['id'] as String,
      userId: json['userId'] as String? ?? json['user_id'] as String? ?? '',
      linkedAccountId: json['linkedAccountId'] as String?,
      linkedAccountName: json['linkedAccountName'] as String?,
      name: json['name'] as String? ?? '',
      targetAmount: (json['targetAmount'] as num?)?.toDouble() ?? 0.0,
      currentAmount: (json['currentAmount'] as num?)?.toDouble() ?? 0.0,
      remainingAmount: (json['remainingAmount'] as num?)?.toDouble() ?? 0.0,
      progressPercentage: (json['progressPercentage'] as num?)?.toDouble() ?? 0.0,
      targetDate: json['targetDate'] != null
          ? DateTime.tryParse(json['targetDate'].toString()) ?? DateTime.now()
          : DateTime.now(),
      daysRemaining: (json['daysRemaining'] as num?)?.toInt() ?? 0,
      paceStatus: GoalPaceStatus.fromString(json['paceStatus'] as String? ?? 'ON_TRACK'),
      expectedPaceAmount: (json['expectedPaceAmount'] as num?)?.toDouble() ?? 0.0,
      paceDifference: (json['paceDifference'] as num?)?.toDouble() ?? 0.0,
      requiredMonthlySavings: (json['requiredMonthlySavings'] as num?)?.toDouble() ?? 0.0,
      requiredWeeklySavings: (json['requiredWeeklySavings'] as num?)?.toDouble() ?? 0.0,
      requiredDailySavings: (json['requiredDailySavings'] as num?)?.toDouble() ?? 0.0,
      isCompleted: json['isCompleted'] as bool? ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }
}

class GoalsSummary {
  final double totalTargetAmount;
  final double totalCurrentAmount;
  final double totalRemainingAmount;
  final double overallProgressPercentage;
  final int activeGoalsCount;
  final int completedGoalsCount;

  const GoalsSummary({
    required this.totalTargetAmount,
    required this.totalCurrentAmount,
    required this.totalRemainingAmount,
    required this.overallProgressPercentage,
    required this.activeGoalsCount,
    required this.completedGoalsCount,
  });

  factory GoalsSummary.fromJson(Map<String, dynamic> json) {
    return GoalsSummary(
      totalTargetAmount: (json['totalTargetAmount'] as num?)?.toDouble() ?? 0.0,
      totalCurrentAmount: (json['totalCurrentAmount'] as num?)?.toDouble() ?? 0.0,
      totalRemainingAmount: (json['totalRemainingAmount'] as num?)?.toDouble() ?? 0.0,
      overallProgressPercentage: (json['overallProgressPercentage'] as num?)?.toDouble() ?? 0.0,
      activeGoalsCount: (json['activeGoalsCount'] as num?)?.toInt() ?? 0,
      completedGoalsCount: (json['completedGoalsCount'] as num?)?.toInt() ?? 0,
    );
  }

  factory GoalsSummary.empty() {
    return const GoalsSummary(
      totalTargetAmount: 0,
      totalCurrentAmount: 0,
      totalRemainingAmount: 0,
      overallProgressPercentage: 0,
      activeGoalsCount: 0,
      completedGoalsCount: 0,
    );
  }
}
